### PR TITLE
sgrep_lint: Support path includes/excludes in rule files

### PR DESCRIPTION
**Pull request by underyx**
Note to reviewers, I recommend checking the documentation file at first to understand the design. I intended to have symmetry between the design of `patterns:` and `paths:` in the YAML definitions.

Closes https://github.com/returntocorp/enterprise/issues/70

### CODE REVIEW

File: docs/config/advanced.md
Comment (dlukeomalley) [inline docs/config/advanced.md:258]: @underyx Thank you for the documentation! Did you consider keys like `exclude` or `include-only`, similar to Zulip's existing terminology and syntax? See https://github.com/zulip/zulip/blob/master/tools/linter_lib/custom_check.py#L7-L13

Similarly, did you consider and decide against accepting a list of paths?

```yaml
paths:
   exclude: ["*.jinja2", "_test.go"]
```

I don't have terribly strong opinion and am just exploring thought process above.

**Comment by dlukeomalley** [comment]
@underyx @DrewDennison Once this merges, can you cut a release and echo back with details on how to externally get it?

File: docs/config/advanced.md
Comment (underyx) [reply docs/config/advanced.md:258]: Yep, I decided against.

- Whichever we choose, customers will have to figure out the intricacies of the config syntax, such as whether items in a list are AND or OR-concatenated. Thought it'd be nice if we save a couple brain cycles for users by just reusing the exact same concepts as for patterns. Although paths is more limited currently and doesn't support stuff like `path-either`, which might be surprising.
- Being able to specify a `filename-not` vs. `path-not` is useful when you don't want to figure out the implementation details of the glob matching, such as how you match all python files in the root of the repo, or all python files in all subdirectories. with the simple exclude list you'd need to know to write `**/*.py`.

With all that said, I'm only 60% sure about this.

**Comment by underyx** [comment]
@DrewDennison can you handle shipping the binary?

File: docs/config/advanced.md
Comment (mschwager) [inline docs/config/advanced.md:248]: Just a heads up, I've also been iterating on the docs here: #476. This will probably require some merging at some point.

Comment (brendongo) [reply docs/config/advanced.md:248]: This is on experimental branch and for now no plans to merge it into develop fwiw @mschwager

File: docs/config/advanced.md
Comment (brendongo) [reply docs/config/advanced.md:258]: fwiw having an exclude or include list is a pretty common pattern in many tools and it's pretty understood that if any file matches an item in an exclude list it is ignored.

Comment (underyx) [reply docs/config/advanced.md:258]: Right, that was a bad example to mention. Should I change the config syntax?

Comment (brendongo) [reply docs/config/advanced.md:258]: but since this is experimental anyway ill leave up to you and @DrewDennison

**Comment by DrewDennison** [comment]
@underyx yes I'll handle shipping the binary with @brendongo but would love to get this merged when you have time. Please flag if you can't get to it today

**Comment by underyx** [comment]
> @underyx yes I'll handle shipping the binary with @brendongo but would love to get this merged when you have time. Please flag if you can't get to it today

I can, though I don't see a clear action here, I'm waiting for an approval or changes requested. I can resolve the conflicts in the meantime.

**Review by DrewDennison — APPROVED** [review]
Code looks fine but haven't tested it myself.

**Comment by underyx** [comment]
cool cool, resolving conflicts now

**Comment by underyx** [comment]
There's tests included that gave the correct results

**Pull request merged**
